plugins/elantp/fu-elantp-hid-haptic-device.c

```diff
@@ -0,0 +1,865 @@
+
+		if (!fu_elantp_hid_haptic_device_send_cmd(FU_DEVICE(parent), blk, blksz, NULL, 0, error))
+			return FALSE;
+			
+		g_usleep(self->fw_page_size == 512 ? ELANTP_DELAY_WRITE_BLOCK_512 * 1000
+						   : ELANTP_DELAY_WRITE_BLOCK * 1000);
+
+		if (!fu_elantp_hid_haptic_device_write_cmd(FU_DEVICE(parent),
+							ETP_CMD_I2C_SET_EEPROM_CTRL,
+							ETP_CMD_I2C_SET_EEPROM_DATATYPE,
+							error))
+				return FALSE;
+
+		ctrl_result = fu_elantp_hid_haptic_device_ensure_eeprom_iap_ctrl(FU_DEVICE(parent),self, error);
+
+		if (ctrl_result == 1) {
+			retry_cnt = 0;	
+
+			/* update progress */
+			checksum += csum_tmp;
+			fu_progress_set_percentage_full(fu_progress_get_child(progress),
+							(gsize)i + 1,
+							(gsize)chunks->len + 1);
+		} else if (ctrl_result == 0) {
+			i -= 1;
+			retry_cnt++;
+			if (retry_cnt >= 3) {
+				g_set_error(error,
+					FWUPD_ERROR,
+					FWUPD_ERROR_WRITE,
+					"bootloader reports failed write: 0x%x",
+					self->iap_ctrl);
+				return FALSE;
+			}
+		} else {
+			g_set_error(error,
+			    	FWUPD_ERROR,
+				FWUPD_ERROR_WRITE,
+				"bootloader reports failed write: 0x%x",
+				self->iap_ctrl);
+			return FALSE;
+		}
+	}
+	fu_progress_step_done(progress);	
+	
+	if (!fu_elantp_hid_haptic_device_write_cmd(FU_DEVICE(parent),
+					    	ETP_CMD_I2C_EEPROM_LONG_TRANS,
+					    	ETP_CMD_I2C_EEPROM_LONG_TRANS_DISABLE,
+					    	error)) {
+		g_prefix_error(error, "cannot disable EEPROM Long Transmission mode: ");
+		return FALSE;
+	}
+	if (!fu_elantp_hid_haptic_device_write_cmd(FU_DEVICE(parent),
+					    	ETP_CMD_I2C_SET_EEPROM_CTRL,
+					    	ETP_CMD_I2C_SET_EEPROM_LEAVE_IAP,
+					    	error)) {
+		g_prefix_error(error, "cannot leave EEPROM IAP: ");
+		return FALSE;
+	}
+	g_usleep(ELANTP_DELAY_RESET * 1000);
+	checksum_device = fu_elantp_hid_haptic_device_get_checksum(FU_DEVICE(parent), error);
+
+	if (checksum != checksum_device) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_WRITE,
+			    "checksum failed 0x%04x != 0x%04x",
+			    checksum,
+			    checksum_device);
+		return FALSE;
+	}
+	
+	fw_ver = fu_elanhaptic_firmware_get_fwver(firmware_elanhaptic);
+	fw_ver_device = fu_elantp_hid_haptic_device_get_version(FU_DEVICE(parent), self, error);
+	if (fw_ver != fw_ver_device) {
+		return FALSE;
+	}
+
+	fu_progress_step_done(progress);
+	return TRUE;
+}
+
+static gboolean
+fu_elantp_hid_haptic_device_detach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuElantpHidDevice *parent = fu_elantp_haptic_device_get_parent(device, error);
+	FuElantpHidHapticDevice *self = FU_ELANTP_HID_HAPTIC_DEVICE(device);
+	guint16 tp_iap_ver;
+	guint16 tp_ic_type;
+	guint8 buf[2] = {0x0};
+	guint16 ctrl; 
+	guint16 tmp;
+	
+	/* sanity check */
+	if (fu_device_has_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER)) 
+		return FALSE;
+
+	if (self->driver_ic != 0x2 || self->iap_ver != 0x1) {
+		g_prefix_error(error, "Can't support this EEPROM IAP: ");
+        	return FALSE;
+   	}
+
+	if (parent == NULL)
```

Comment:
I think move the fu_elantp_haptic_device_get_parent here, not in the initializer block as it can set GError